### PR TITLE
Support nav_msgs/Path in Webviz

### PR DESCRIPTION
**Pull request by cjds**
## Summary

Add support for the message nav_msgs/Path to webviz
## Test plan

I built locally and tested with a bag with a Path in it. Here's a gif of the Path  in Webviz

## Versioning impact

Might recommend a patch version at most

### CODE REVIEW

**Comment by cjds** [comment]
Also seems like there's an issue with the test here. Not sure how to fix

```
{ CredentialsError: Missing credentials in config
    at IncomingMessage.<anonymous> (/home/circleci/project/node_modules/aws-sdk/lib/util.js:865:34)
    at IncomingMessage.emit (events.js:203:15)
    at IncomingMessage.EventEmitter.emit (domain.js:448:20)
    at endReadableNT (_stream_readable.js:1145:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
  message: 'Missing credentials in config',
  retryable: false,
  time: 2021-03-04T21:46:00.662Z,
  code: 'CredentialsError',
  originalError:
   { message: 'Could not load credentials from any providers',
     retryable: false,
     time: 2021-03-04T21:46:00.662Z,
     code: 'CredentialsError' } }
```

File: packages/webviz-core/src/panels/ThreeDimensionalViz/SceneBuilder/index.js [hunk 1]
```diff
@@ -796,6 +798,25 @@ export default class SceneBuilder implements MarkerProvider {
...
+        const newMessage = {
+          header: deepParse(pathStamped.header()),
+          // TODO(@cjds) Make this make use of the orientation of the poses in the path as well
+          points: pathStamped
+            .poses()
+            .toArray()
+            .map((pose) => deepParse(pose.pose().position())),
+          closed: false,
+          scale: { x: 0.2 },
+          color: { r: 1, g: 0, b: 0, a: 1 },
```
Comment (vidaaudrey) [inline packages/webviz-core/src/panels/ThreeDimensionalViz/SceneBuilder/index.js:815]: Would be nice to add `overrideColor` topic settings for setting r,g,b,a. Similar to rviz: http://wiki.ros.org/rviz/DisplayTypes/Path

File: packages/webviz-core/src/panels/ThreeDimensionalViz/SceneBuilder/index.js [hunk 2]
```diff
@@ -796,6 +798,25 @@ export default class SceneBuilder implements MarkerProvider {
...
         this._consumeOccupancyGrid(topic, deepParse(message));
         break;
+      case NAV_MSGS_PATH_DATATYPE: {
+        const pathStamped = cast<BinaryPath>(message);
+        if (pathStamped.poses().length() === 0) {
+          break;
+        }
+        const newMessage = {
+          header: deepParse(pathStamped.header()),
+          // TODO(@cjds) Make this make use of the orientation of the poses in the path as well
```
Comment (vidaaudrey) [inline packages/webviz-core/src/panels/ThreeDimensionalViz/SceneBuilder/index.js:808]: Curious what do you have in mind for orientation visualization? Some sort of arrow?

**Review by vidaaudrey — COMMENTED** [review]
Thanks for the contribution. Would be nice if you could add a story for testing. Can refer this component: https://github.com/cruise-automation/webviz/blob/master/packages/webviz-core/src/panels/ThreeDimensionalViz/stories/MarkerStory.js. If you have some sample data handy, I can help you add one. To get sample data, you can use the Internal panel and click `Record raw data`:
image

**Comment by cjds** [comment]
I'm a little busy this week. Going to work on this over the weekend

File: packages/webviz-core/src/panels/ThreeDimensionalViz/SceneBuilder/index.js (hunk 1 above)
Comment (cjds) [reply packages/webviz-core/src/panels/ThreeDimensionalViz/SceneBuilder/index.js:815]: That would be awesome. Will change it. I copied this from `PolygonStamped` .. Was there any reason that didn't use the overrides?
https://github.com/cruise-automation/webviz/blob/eab0da004a6bc6e90a9107d1c46a0df3337cf70e/packages/webviz-core/src/panels/ThreeDimensionalViz/SceneBuilder/index.js#L838

**Comment by cjds** [comment]
@vidaaudrey added sample data here https://drive.google.com/file/d/1xEoG3L6u5tPVwijMh_A1VnzSvQTCVpSB/view?usp=sharing

Still a little confused about how to make a story though? I see this using the fixture but I don't see how to use it?

File: packages/webviz-core/src/panels/ThreeDimensionalViz/SceneBuilder/index.js (hunk 1 above)
Comment (cjds) [reply packages/webviz-core/src/panels/ThreeDimensionalViz/SceneBuilder/index.js:817]: wv2
Got the override working

**Review by vidaaudrey — APPROVED** [review]
LGTM. Since we need to add the support internally as well, I'll take the data to add the story internally. I've created a PR and sent to our team to review. You may expect it to show up in the open source version next week or so.

**Comment by cjds** [comment]
Good to know. thanks for the update. Should I close the PR when it makes it in in the open source build?

**Comment by vidaaudrey** [comment]
Yeah, you could close it later. I will tag you.

On Mar 18, 2021, at 6:29 PM, Carl Saldanha ***@***.***> wrote:

﻿

Good to know. thanks for the update. Should I close the PR when it makes it
in in the open source build?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/cruise-automation/webviz/pull/592#issuecomment-802443905>,
or unsubscribe

**Comment by vidaaudrey** [comment]
@cjds FYI this PR [here](https://github.com/cruise-automation/webviz/pull/606/files#diff-7513b2529050b577a79ce892656054b86ed5b235ecbcf41e451cf11414bebf10R776-R794) added the support for `nav_msgs/Path`. Will close this PR now. Let me know if you have any questions.

**Comment by yyu233** [comment]
webviz_path_msg_not_working
I pulled the docker image from https://hub.docker.com/r/cruise/webviz
It doesn't draw anything on the 3D plot.
Also, I ran into this issue once:
webviz_path_msg_not_working_2

**Comment by jtbandes** [comment]
Try [Foxglove Studio](https://foxglove.dev/ros), which supports nav_msgs/Path: https://foxglove.dev/docs/studio/panels/3d#path

**Pull request closed without merging**